compiler/rustc_parse/src/parser/item.rs

```diff
@@ -2278,6 +2278,18 @@ impl<'a> Parser<'a> {
                     err.span_label(ident.span, "while parsing this `fn`");
                     err.emit();
                 } else {
+                    // check for typo'd Fn* trait bounds such as
+                    // fn foo<F>() where F: FnOnce -> () {}
+                    if self.token.kind == token::RArrow
+                        && let Some(t) = [sym::FnOnce, sym::FnMut, sym::Fn]
+                            .into_iter()
+                            .find(|s| self.prev_token.is_ident_named(*s))
+                    {
```

Comment:
```suggestion
                    if self.token.kind == token::RArrow {
```

You could theoretically generalize this by not special-casing `Fn{,Mut,Once}` similar to #104531 (#108185). This would allow us to suggest adding `()` in cases like

```rs
fn f(_: impl Function -> ()) {}
use FnOnce as Function;
```

Reply:
What would be a good message in that case? Maybe something like
diagnostic: "expected a function body, found `->`"
note: "you may be trying to write a `Fn*` family trait bound, which require arguments in parentheses"
suggestion: (maybe-incorrect) "try adding parentheses"

Reply:
Hmm, I'd just say “`Fn` bounds require arguments in parentheses” or “expected list of arguments before `->`”, not sure. #104531 just says “`Fn` traits […]” showing complete disregard for the concrete trait (`FnOnce`, `FnMut`, …).

Reply:
Personally I'd prefer “`Fn`-family (trait) bound” but I'm not sure if we already use this terminology already in diagnostics. “expected a function *body* [my emphasis], found ->” wouldn't be correct, since it's about the argument list, not a body.

And yeah it's maybe-incorrect, I guess. You could make it conditional on the name. So if it isn't `Fn{,Mut,Once}` the suggestion is maybe-incorrect, otherwise, machine-applicable.